src/quote/provider/icusd.ts

```diff
@@ -0,0 +1,151 @@
+import { BigNumber } from '@ethersproject/bignumber'
+import { Address } from 'viem'
+
+import { TheUSDCYieldIndex } from 'constants/tokens'
+import {
+  FlashMintNavTransactionBuilder,
+  FlashMintWrappedBuildRequest,
+  WrappedTransactionBuilder,
+} from 'flashmint'
+import { getExpectedReserveRedeemQuantity } from 'utils/custom-oracle-nav-issuance-module'
+import { getUsdcBalance } from 'utils/erc20'
+
+import { FlashMintNavQuoteProvider } from '../flashmint/nav'
+import { WrappedQuoteProvider } from '../flashmint/wrapped'
+import { QuoteProvider, QuoteToken } from '../interfaces'
+import {
+  FlashMintContractType,
+  FlashMintQuote,
+  FlashMintQuoteRequest,
+} from '../provider'
+import { SwapQuoteProvider } from '../swap'
+import { buildQuoteResponse } from './utils'
+
+export interface IcUsdQuoteRequest extends FlashMintQuoteRequest {
+  chainId: number
+  isMinting: boolean
+  inputToken: QuoteToken
+  outputToken: QuoteToken
+  indexTokenAmount: BigNumber
+  // FlashMintNav always needs the input amount
+  inputTokenAmount: BigNumber
+  slippage: number
+}
+
+export class IcUsdQuoteRouter
+  implements QuoteProvider<IcUsdQuoteRequest, FlashMintQuote>
+{
+  icUsd = TheUSDCYieldIndex
+  constructor(
+    private readonly rpcUrl: string,
+    private readonly swapQuoteProvider: SwapQuoteProvider
+  ) {}
+
+  async getQuote(request: IcUsdQuoteRequest): Promise<FlashMintQuote | null> {
+    // TODO: remove in production; for testing only
+    if (request.isMinting) {
+      return await this.getFlashMintWrappedQuote(request)
+    }
+    if (request.isMinting) {
+      return await this.getFlashMintNavQuote(request)
+    } else {
+      const { chainId, indexTokenAmount, inputTokenAmount } = request
+      const usdcBalance = await getUsdcBalance(
+        '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
+        this.icUsd.address as Address,
+        chainId
+      )
+      const usdcInputAmount = await getExpectedReserveRedeemQuantity(
+        this.icUsd.address as Address,
+        '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
+        indexTokenAmount.toBigInt()
+      )
+      console.log(usdcBalance.toString(), 'USDC')
+      // 80% of the USDC balance of icUSD
+      const threshold = (usdcBalance * 80n) / 100n
+      const useFlashMintNav = usdcInputAmount < threshold
```

Comment:
@edkim @pblivin0x please check if this threshold logic when to use FMNav or FMWrapped for redeeming is correct.